rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,739 @@
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+
+### 4.2.1.2 Migration algorithm: Cloned index per version
+Note:
+- The description below assumes the migration algorithm is released in 7.10.0.
+  So < 7.10.0 will use `.kibana` and >= 7.10.0 will use `.kibana_current`.
+- We refer to the alias and index that outdated nodes use as the source alias
+  and source index.
+- Every version performs a migration even if mappings or documents aren't outdated.
+
+1. Locate the source index by fetching aliases (including `.kibana` for
+   versions prior to v7.10.0)
+   
+   ```
+   GET '/_alias/.kibana_current,.kibana_7.10.0,.kibana'
+   ```
+   
+   The source index is:
+   1. the index the `.kibana_current` alias points to, or if it doesn’t exist, 
+   2. the index the `.kibana` alias points to, or if it doesn't exist,
+   3. the v6.x `.kibana` index
+2. If `.kibana_current` and `.kibana_7.10.0` both exists and are pointing to the same index this version's migration has already been completed.
+   1. Because the same version can have plugins enabled at any point in time, perform the mappings update in step (6).
+   2. Skip to step (9) to start serving traffic.
+3. Fail the migration if:
+   1. `.kibana_current` is pointing to an index that belongs to a later version of Kibana .e.g. `.kibana_7.12.0_001`
+   2. The source index contains documents that belong to an unknown Saved Object type (from a disabled plugin). Log an error explaining that the plugin that created these documents needs to be enabled again or that these objects should be deleted. See section (4.2.1.4).
+4. Mark the source index as read-only and wait for all in-flight operations to drain (requires https://github.com/elastic/elasticsearch/pull/58094). This prevents any further writes from outdated nodes. Assuming this API is similar to the existing `/<index>/_close` API, we expect to receive `"acknowledged" : true` and `"shards_acknowledged" : true`. If all shards don’t acknowledge within the timeout, retry the operation until it succeeds.
+5. Clone the source index into a new target index which has writes enabled. All nodes on the same version will use the same fixed index name e.g. `.kibana_7.10.0_001`. The `001` postfix isn't used by Kibana, but allows for re-indexing an index should this be required by an Elasticsearch upgrade. E.g. re-index `.kibana_7.10.0_001` into `.kibana_7.10.0_002` and point the `.kibana_7.10.0` alias to `.kibana_7.10.0_002`.
```

Comment:
As far as I'm aware, cloning indices in Cloud will not use hard-links until https://github.com/elastic/elasticsearch/issues/61145 is implemented, so we'll be consuming excess disk space in the meantime.

Reply:
I wasn't aware of that limitation 😭 

We don't have any data on the size of Kibana indices, although we could collect some from 7.10 as part of https://github.com/elastic/kibana/issues/56762

We have always created duplicate indices, so cloning without hard-link support will be similar to what we have today. The difference is that today Kibana typically migrates in minors but not in bug fix releases whereas the proposed algorithm will create a new index for every bug fix too. However, users with large kibana indices would have always had to monitor and clean up their indices, so even though this might accelerate it it shouldn't be totally unexpected that Kibana requires more disk space after an upgrade.

I'm also following-up with Cloud/ES as it seems like it might be a quick fix to get hard-link support.

But I'm leaning towards just continuing with the current design. Operationally having an index per version makes it very easy to delete old indices without making any mistakes, so even if it takes up more data it's easier and safer to clear up that data.

Reply:
> We have always created duplicate indices, so cloning without hard-link support will be similar to what we have today. The difference is that today Kibana typically migrates in minors but not in bug fix releases whereas the proposed algorithm will create a new index for every bug fix too. However, users with large kibana indices would have always had to monitor and clean up their indices, so even though this might accelerate it it shouldn't be totally unexpected that Kibana requires more disk space after an upgrade.

> I'm also following-up with Cloud/ES as it seems like it might be a quick fix to get hard-link support.

> But I'm leaning towards just continuing with the current design. Operationally having an index per version makes it very easy to delete old indices without making any mistakes, so even if it takes up more data it's easier and safer to clear up that data.

Seems reasonable to me. If we don't plan on fixing this in Cloud for a super long time, it potentially becomes a larger issues. But, as long as we fix in the next few minors, the impact seems negligible.